### DISCUSSION TITLE
Mac security Handling in OpenThread

### DISCUSSION THREAD

**Discussion by joseph-jayanth** [openthread/openthread#10586] [Q&A] [answered]
Hi Members 

In RCP model of open thread implementation where does the mac security can be handled? 
Should it be done in the Open thread controller or in the open-thread core(host processor) in the host?
If it can be done in the controller should it be done in HW crypto engine or in between the spinel and OT drivers? 
Also please advise on the config aspects related to the mac security handling for Open thread!

Thanks in advance.

Regards
Joseph
![image (1)](https://github.com/user-attachments/assets/1aeccc48-7363-4585-b535-05735c873dcd)

**Comment by abtink** [comment, marked as answer]
> In RCP model of open thread implementation where does the mac security can be handled?
> Should it be done in the Open thread controller or in the open-thread core(host processor) in the host?

The RCP model requires the RCP to provide and handle TX frame security processing (for Thread 1.2 or later versions). RX frame security processing is done on the host in the `Mac` layer.

Regarding the firmware running on the RCP itself, TX security processing can be handled by either OpenThread's `Mac/SubMac` implementation or delegated to the radio platform. The radio platform indicates its capabilities (see `OT_RADIO_CAPS`), and the OT stack will delegate supported functionality to the platform or use its own implementation otherwise. The choice is up to the radio platform, but typically they provide `OT_RADIO_CAPS_TRANSMIT_SEC`.

https://github.com/openthread/openthread/blob/19dadd9c4524700a0747e1f40e6e1f36b80c9ce9/include/openthread/platform/radio.h#L124-L141

-----

> If it can be done in the controller should it be done in HW crypto engine or in between the spinel and OT drivers?

This decision is entirely up to the platform.

* If TX frame security processing is implemented in the radio platform, the platform developer decides how to handle it.
* When security processing is performed by the OpenThread stack, OT defines a set of platform APIs [`otPlatCrypto`](https://github.com/openthread/openthread/blob/19dadd9c4524700a0747e1f40e6e1f36b80c9ce9/include/openthread/platform/crypto.h#L380-L433), allowing the platform to specify its crypto implementation. A default implementation using mbedtls is provided by the OT stack in [`crypto_platform.cpp`](https://github.com/openthread/openthread/blob/19dadd9c4524700a0747e1f40e6e1f36b80c9ce9/src/core/crypto/crypto_platform.cpp
), but the platform can choose to provide its own implementation (which may use HW crypto engines).

**Reply by jchellap** [reply]
Thanks @abtink 
What about Secured RX frame processing in RCP model? Can it be passed onto OT submac for decryption rather than platform level processing? What is the suggestions/recommendation? Should open thread controller handle it or handled by open thread core?

**Reply by abtink** [reply]
RX security processing is exclusively handled by the OT core (within the `Mac` class). The radio platform and/or RCP are not involved in this process.

**Reply by jchellap** [reply]
Thanks, that clarifies.

**Reply by joseph-jayanth** [reply]
Thanks @abtink for the clarification!
